File: web-sdk/src/domain/models/command/custom-command.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

// Channel -> Client custom commands.
//
// The channel can ask the client to run a named command by sending a custom
// command request. The host registers a handler per command id. The handler
// receives the request payload and returns the response payload. Throwing (or
// rejecting) reports the command as failed back to the channel.
export type CustomCommandHandler = (
  payload: string
) => string | void | Promise<string | void>;

// Outcome reported back to the channel for a custom command request.
export type CustomCommandStatus = 'success' | 'error';

// A custom command request received from the channel.
export interface CustomCommandInvocation {
  commandId: string;
  payload: string;
  // Echoed back on the response so the channel can correlate it.
  correlationId: string;
}

File: flutter-sdk/lib/domain/models/command/custom_command.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:async';

// Handler the host registers for a custom command, keyed by the command id the
// channel sends in a CustomCommandRequest. The handler receives the request
// payload and returns the response payload the SDK sends back (filling in the
// status, timestamp and correlation id). It may be synchronous or asynchronous.
// Returning null sends an empty payload. Throwing reports the command as failed
// to the channel.
typedef CustomCommandCallback = FutureOr<String?> Function(String payload);
